### PR TITLE
perf: use `finishModules`

### LINKED ISSUE #70: The same file is being linted multiple times during a compilation, causing very long build times
- Operating System: macOS 11.1
- Node Version: 15.6.0
- NPM Version: yarn berry using PnP
- webpack Version: 4.46.0
- eslint-webpack-plugin Version: 2.4.3

### Expected Behavior

Each module file should be linted only once.

### Actual Behavior

The same module file is being linted multiple times unnecessarily which drastically increases the build time.

### Code

```js
// webpack.config.js
new ESLintPlugin({
  threads: true,
  extensions: ['js', 'vue'],
})
```

### How Do We Reproduce?

The same module file will be linted each time it is required with a different resource query string. This code strips off the query string:

https://github.com/webpack-contrib/eslint-webpack-plugin/blob/f819203302d26a1beadb66ca42270bfa92c65927/src/index.js#L82-L95

In my case I'm using `vue-loader` which loads `.vue` files. It does some trickery behind the scenes such that when you require a `.vue` module it will translate that to multiple requires to load all the different parts of the file:

```js
// This...
import Comp from './comp.vue';

// ...is converted into something like this
import Comp from './comp.vue?type=js';
import './comp.vue?type=css';
```

`eslint-webpack-plugin` lints the `comp.vue` file twice in this case. I did some logging and some of my modules are being linted 5 times, maybe more.

I did a quick test with this modification:

```js
const seen = new Set();

 const processModule = (module) => {
   if (module.resource) {
     const file = module.resource.split('?')[0];
  
     if (
       file &&
       !seen.has(file) &&
       micromatch.isMatch(file, wanted) &&
       !micromatch.isMatch(file, exclude)
     ) { 
       // Queue file for linting.
       seen.add(file);
       lint(file);
     }
  }
};
```

and my initial build time reduced from 300s (!) down to 100s.

If I run `yarn eslint .` in my project directory it takes 44s but it seems `eslint-webpack-plugin` takes much longer to do essentially the same thing.

I tried setting the `cache` option to true (if such an option exists), thinking it would skip linting files that have already been linted so far in the compilation, but that made no difference.

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Use the hook `finishModules` to lint more files and not one by one

Resolve #70

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Let's add test, should be not hard, because we can spy `eslint`

**Review by alexander-akait — APPROVED** [review]

**Comment by watjurk** [comment]
I think that using "outside set" and checking if file is in this set is kind of performance issue, as finishModules hook is called only once we can came with better solution:

```js
const resources = Array.from(modules)
          // @ts-ignore
          .filter((module) => module.resource)
          // @ts-ignore
          .map(({ resource }) => resource.split('?')[0])
          .filter((resource) => resource /* && !seen.has(resource) */ && isMatch(resource, wanted) && !isMatch(resource, exclude));
const resourcesUnique = new Set(resources);
//...
```

File: src/index.js [hunk 1]
```diff
@@ -95,27 +95,24 @@ class ESLintWebpackPlugin {
...
+            .filter((module) => module.resource)
+            // @ts-ignore
+            .map(({ resource }) => resource.split('?')[0])
+            .filter(
+              (resource) =>
+                resource &&
+                isMatch(resource, wanted) &&
+                !isMatch(resource, exclude)
+            )
+        );
```
Comment (alexander-akait) [inline src/index.js:112]: Recreating `modules` from `Array` and when convert it to `Set` is big perf problem, just is `for (const module of modules) { }` fir the best perf

Comment (ricardogobbosouza) [reply src/index.js]: Updated using spread operator

Comment (ricardogobbosouza) [reply src/index.js]: WDTY? https://github.com/webpack-contrib/eslint-webpack-plugin/pull/73/commits/79d6aba24c0332f98904c28ea03da77f92ac6db0#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556

**Comment by jsg2021** [comment]
By moving the finishModule hook we no longer spread the linting work out... one lump (in my case 5620 files) gets sent to the linter in one go. The thread pool (if threaded) does not fill up... only one worker gets all the jobs.

**Pull request merged**